src/domain.c

```diff
@@ -157,19 +180,43 @@ ld_config_t *ld_load_config(TALLOC_CTX* ctx, const char *filename)
     get_config_optional_string("cert_file", certfile);
     get_config_optional_string("key_file", keyfile);
 
-    result->cacertfile = cacertfile
-            ? talloc_strndup(ctx, cacertfile, strlen(cacertfile))
-            : talloc_strndup(ctx, empty_string, strlen(empty_string));
-    result->certfile = certfile
-            ? talloc_strndup(ctx, certfile, strlen(certfile))
-            : talloc_strndup(ctx, empty_string, strlen(empty_string));
-    result->keyfile = keyfile
-            ? talloc_strndup(ctx, keyfile, strlen(keyfile))
-            : talloc_strndup(ctx, empty_string, strlen(empty_string));
+    if (cacertfile)
+    {
+        ld_talloc_strndup(result->cacertfile, error_exit, result, cacertfile, strlen(cacertfile))
+    }
+    else
+    {
+        ld_talloc_strndup(result->cacertfile, error_exit, result, empty_string, strlen(empty_string))
+    }
+
+    if (certfile)
+    {
+        ld_talloc_strndup(result->certfile, error_exit, result, certfile, strlen(certfile))
+    }
+    else
+    {
+        ld_talloc_strndup(result->certfile, error_exit, result, empty_string, strlen(empty_string))
+    }
+
+    if (keyfile)
+    {
+        ld_talloc_strndup(result->keyfile, error_exit, result, keyfile, strlen(keyfile))
+    }
+    else
+    {
+        ld_talloc_strndup(result->keyfile, error_exit, result, empty_string, strlen(empty_string))
+    }
 
     config_destroy(&cfg);
 
     return result;
+
+    error_exit:
+        if (result)
+        {
+            ld_talloc_free(result, error_exit);
```

Comment:
Do not use `error_exit` mark in `error_exit` statement it will result in the endless cycle in case if `talloc_free` fails. `talloc_free` - returns 0 on success and -1 on error. A possible failure condition is if the pointer had a destructor attached to it and the destructor returned -1. Likewise, if "ptr" is NULL, then the function will make no modifications and returns -1. And we in `LD_FREE_HELPER` make `ctx` `NULL` this will result in endless cycle of failures.